enhancements/cloud-integration/tokenized-auth-enablement-operators-on-cloud.md

```diff
@@ -0,0 +1,404 @@
+8. operator loads Secret, makes cloud service requests
+
+
+For the operator author team:
+- Add code to create a CredentialRequests (it will not be stored in the bundle 
+as part of this EP)
+- Add "role-arn"-type fields to the CredentialRequests
+- Add eventing to report status on a CR to indicate lacking STS credentials for fully operational deploy or update.
+
+For the Cloud Credential Operator:
+- Allow detection that a cluster is using time-based tokens and create a Secret containing the ARN + token path per 
+ cloud platform.
+
+For Operator Administrator:
+- Supply cloud credentials for STS (ARN ID, etc) in the known location.
+
+For OperatorHub, Console:
+- Prompt for missing credentials that prevent fully operational install on STS cluster. Same for upgrades.
+- Badge indicating "STS enabled" (this is a naming suggestion to avoid implying that non-badged operators not needing
+ cloud provider resource access would not work) operators.
+
+This graph shows the system as it needs to work (AWS only), not the proposed solution workflow:
+```mermaid
+graph  LR
+  sa([ServiceAccount Signing Keys]) -- Public <br> Key -->s3[S3 Bucket w/ OIDC Config];
+  sa([ServiceAccount Signing Keys]) -- Private <br> Key -->token_signing;
+  service<-.->config
+  service<-.->Pod
+  Pod-->role
+  subgraph cluster [AWS]
+  s3;
+  iam_id[IAM Identity Provider] -- Trust Tokens<br>Signed -->s3;
+  role[IAM Role For Pod]<-->service[AWS Services];
+  end
+  subgraph cluster2 [OpenShift]
+  token_signing[Token Signing]-->projected[Projected<br>ServiceAccount<br>Token];
+  subgraph Pod
+  config[AWS Config File]-->projected
+  end
+  end
+  classDef plain fill:#ddd,stroke:#fff,stroke-width:4px,color:#000;
+  classDef k8s fill:#326ce5,stroke:#fff,stroke-width:4px,color:#fff;
+  classDef cluster fill:#fff,stroke:#bbb,stroke-width:2px,color:#326ce5;
+  classDef cluster2 fill:#fff,stroke:#bbb,stroke-width:2px,color:#326ce5;
+  class config,role,service,s3,projected,token_signing,iam_id,pod1,pod2 k8s;
+  class sa plain;
+  class cluster cluster;
+  class cluster2 cluster1;
+```
+
+### API Extensions
+
+Added to the CredentialsRequest API 
+
+We outline this change using the AWS specific change, but other cloud provider specs will follow a similar pattern:
+Adding to the `spec.ProviderSpec` field ([ref](https://github.com/openshift/cloud-credential-operator/blob/38e7f96bf73036924abcb1352222c17704e4b2e0/pkg/apis/cloudcredential/v1/types_credentialsrequest.go#L53-L55)),<br>
+where `spec.ProviderSpec` would be an `AWSProviderSpec` ([ref](https://github.com/openshift/cloud-credential-operator/blob/38e7f96bf73036924abcb1352222c17704e4b2e0/pkg/apis/cloudcredential/v1/types_aws.go#L27-L31)) for AWS-directed TAT.<p>
+CredentialsRequests:
+```go
+type CredentialsRequestSpec struct {
+...
+
+// ProviderSpec contains the cloud provider specific credentials specification.
+// +kubebuilder:pruning:PreserveUnknownFields
+ProviderSpec *runtime.RawExtension `json:"providerSpec,omitempty"`
```

Comment:
As someone who maintains an API that looks like this, I would strongly advise against using a RawExtension.

Instead, created a discriminated union, that will look something like

```
credentialsRequest:
  platform: AWS
  aws:
    ststIAMRoleARN
```
You still have the `ProviderSpec` for each of your platforms, but they are direct members of the `CredentialsRequestSpec` and you can use a discriminator (`platform`) to define which of the platforms it should be looking at with the union

Please rework the structs into the DU pattern and I'll continue the review from there

Reply:
@JoelSpeed `RawExtension` is unfortunately [already being used](https://github.com/openshift/cloud-credential-operator/blob/38e7f96bf73036924abcb1352222c17704e4b2e0/pkg/apis/cloudcredential/v1/types_credentialsrequest.go#L53-L55) for platform specific `ProviderSpec`.

Reply:
I think the code refs to ProviderSpec and AWSProviderSpec is sufficient to answer this concern, no?

Reply:
Maybe this suggestion helps?
https://github.com/openshift/enhancements/pull/1339/files#r1222378967

Reply:
Yeah the suggestion helps clarify this for me, thanks